File: 2018/06-Jun/6-23-Bit-Field/CSharp/BitField.cs
/**
 * Demonstrate bit fields in C# with enums
 * Author: Andrew Jarombek
 * Date: 1/15/2019
 */

using System;
using static System.Diagnostics.Debug;

namespace BitFields
{
    class BitField
    {
        
        /// <summary>
        /// The [Flags] attribute should always be used when the members of an enum are combinable.
        /// Combinable enums can be used as a bit field.
        /// </summary>
        [Flags]
        private enum User
        {
            // Different statuses of a user
            None = 0,
            Validated = 1,
            Subscribed = 2,
            Admin = 4,
            
            // Combination members can be included in the enum declaration
            ValidatedAdmin = Validated | Admin,
            SubscribedAdmin = Subscribed | Admin,
            ValidatedSubscribed = Validated | Subscribed,
            All = ValidatedSubscribed | Admin
        }

        /// <summary>
        /// Add new fields to an enum using a bit mask
        /// </summary>
        /// <param name="target">A user enum to apply a bit mask to</param>
        /// <param name="mask">A bit mask to apply to the bit field (User enum)</param>
        /// <returns>A new User enum</returns>
        private static User Add(User target, User mask)
        {
            return target | mask;
        }

        /// <summary>
        /// Toggle fields on and off in an enum using a bit mask
        /// </summary>
        /// <param name="target">A user enum to apply a bit mask to</param>
        /// <param name="mask">A bit mask to apply to the bit field (User enum)</param>
        /// <returns>A new User enum</returns>
        private static User Toggle(User target, User mask)
        {
            return target ^ mask;
        }

        /// <summary>
        /// Remove fields from an enum using a bit mask
        /// </summary>
        /// <param name="target">A user enum to apply a bit mask to</param>
        /// <param name="mask">A bit mask to apply to the bit field (User enum)</param>
        /// <returns>A new User enum</returns>
        private static User Remove(User target, User mask)
        {
            return target & ~mask;
        }

        /// <summary>
        /// Check if fields exist in an enum using a bit mask
        /// </summary>
        /// <param name="target">A user enum to apply a bit mask to</param>
        /// <param name="mask">A bit mask to apply to the bit field (User enum)</param>
        /// <returns>true if the bit field (User enum) contains the mask, false otherwise</returns>
        private static bool Contains(User target, User mask)
        {
            return (target & mask) > 0;
        }
        
        static void Main(string[] args)
        {
            const User user = BitField.User.None;
            
            // Adding the Admin field to the None field creates an admin
            Assert(BitField.Add(user, User.Admin) == User.Admin);
            
            // Adding the Subscribed field to the None field creates a subscribed user.  This also proves that 
            // add() doesn't mutate the 'user' value.
            Assert(BitField.Add(user, User.Subscribed) == User.Subscribed);

            const User user2 = BitField.User.ValidatedSubscribed;
            
            // Prove that an enum with combined flags contains both flags individually.
            Assert(BitField.Contains(user2, User.Validated));
            Assert(BitField.Contains(user2, User.Subscribed));
            
            Assert(BitField.Remove(user2, User.Subscribed) == User.Validated);
            
            // Toggle the Admin field on, then toggle the Validated field off
            Assert(BitField.Toggle(user2, User.Admin) == User.All);
            Assert(BitField.Toggle(user2, User.Validated) == User.Subscribed);
        }
    }
}

File: 2018/06-Jun/6-23-Bit-Field/C/bitfield.c
/*
 * Author: Andrew Jarombek
 * Date: 6/22/2018
 * Demonstrate bit fields in C
 */

#define VALIDATED 01
#define SUBSCRIBED 02
#define ADMIN 04

#include <string.h>
#include <stdlib.h>
#include <stdio.h>

// Type to hold a user construct
typedef struct {
    char first[31];
    char last[31];
    unsigned int statusField : 3;
} User;

void addStatus(User *user, unsigned int mask);
void removeStatus(User *user, unsigned int mask);
int containsStatus(unsigned int bitField, unsigned int mask);
void printUser(User user);

int main()
{
    User andy;
    strcpy(andy.first, "Andrew");
    strcpy(andy.last, "Jarombek");

    // Set the status as 000
    andy.statusField = 0;

    printUser(andy);

    // Change the status from 000 to 001
    addStatus(&andy, VALIDATED);

    printUser(andy);

    // Change the status from 001 to 101
    addStatus(&andy, ADMIN);

    // The user now is validated and is an admin
    printUser(andy);

    // Take away the admin rights to the user, change the status from 101 to 001
    removeStatus(&andy, ADMIN);
    printUser(andy);

    // Give the subscribed status, change the status from 001 to 011
    addStatus(&andy, SUBSCRIBED);
    printUser(andy);

    if (containsStatus(andy.statusField, VALIDATED)) {
        printf("The User IS Validated \n");
    } else {
        printf("The User IS NOT Validated \n");
    }

    if (containsStatus(andy.statusField, SUBSCRIBED)) {
        printf("The User IS Subscribed \n");
    } else {
        printf("The User IS NOT Subscribed \n");
    }

    if (containsStatus(andy.statusField, ADMIN)) {
        printf("The User IS an Admin \n");
    } else {
        printf("The User IS NOT an Admin \n");
    }
}

/**
 * Add a status to the bit field on a user type
 * @param user - a User type
 * @param mask - the status to add to the bit field
 */
void addStatus(User *user, unsigned int mask) {
    // Equivalent of statusField = statusField | mask.  Combines the "on" bits
    // from the existing statuses on the bit field and the new status bits
    user->statusField |= mask;
}

/**
 * Remove a status from the bit field on a user type
 * @param user - a User type
 * @param mask - the status to add to the bit field
 */
void removeStatus(User *user, unsigned int mask) {
    // Equivalent of statusField = statusField & ~mask.
    user->statusField &= ~mask;
}

/**
 * Check to see if a status is in the bit field
 * @param bitField - a bit field to search through
 * @param mask - a bit mask to bitwise 'and' with the bit field
 * @return 0 if the mask doesn't exist in the bit field, an integer >= 1 otherwise
 */
int containsStatus(unsigned int bitField, unsigned int mask) {
    return bitField & mask;
}

/**
 * Print out a User types properties
 * @param user - a User type
 */
void printUser(User user) {
    printf("%s, %s, %d \n", user.first, user.last, user.statusField);
}